File: sdk/python/src/cq/scoring.py
"""Confidence scoring and relevance functions for knowledge units."""

from datetime import UTC, datetime

from ._util import _as_list
from .models import Flag, FlagReason, KnowledgeUnit

CONFIRMATION_BOOST = 0.1
FLAG_PENALTY = 0.15
_CONFIDENCE_CEILING = 1.0
_CONFIDENCE_FLOOR = 0.0

_RELEVANCE_CEILING = 1.0
_RELEVANCE_FLOOR = 0.0

_DOMAIN_WEIGHT = 0.55
_LANGUAGE_WEIGHT = 0.15
_FRAMEWORK_WEIGHT = 0.15
_PATTERN_WEIGHT = 0.15


def apply_confirmation(unit: KnowledgeUnit) -> KnowledgeUnit:
    """Increment confirmations and boost confidence, capped at 1.0."""
    new_confidence = min(unit.evidence.confidence + CONFIRMATION_BOOST, _CONFIDENCE_CEILING)
    new_confirmations = unit.evidence.confirmations + 1
    return unit.model_copy(
        update={
            "evidence": unit.evidence.model_copy(
                update={
                    "confidence": new_confidence,
                    "confirmations": new_confirmations,
                    "last_confirmed": datetime.now(UTC),
                }
            )
        }
    )


def apply_flag(
    unit: KnowledgeUnit,
    reason: FlagReason,
    *,
    duplicate_of: str | None = None,
) -> KnowledgeUnit:
    """Reduce confidence and record the flag reason."""
    new_confidence = max(unit.evidence.confidence - FLAG_PENALTY, _CONFIDENCE_FLOOR)
    new_flag = Flag(reason=reason, duplicate_of=duplicate_of)
    return unit.model_copy(
        update={
            "evidence": unit.evidence.model_copy(update={"confidence": new_confidence}),
            "flags": [*unit.flags, new_flag],
        }
    )


def calculate_relevance(
    unit: KnowledgeUnit,
    query_domains: list[str],
    query_languages: list[str] | None = None,
    query_frameworks: list[str] | None = None,
    query_pattern: str = "",
) -> float:
    """Score relevance from 0.0 to 1.0 based on domain overlap and context match.

    Domain overlap is the primary signal (weighted at 0.55).
    Language, framework, and pattern matches are secondary signals (0.15 each).
    """
    query_domains = _as_list(query_domains)
    if query_languages is not None:
        query_languages = _as_list(query_languages)
    if query_frameworks is not None:
        query_frameworks = _as_list(query_frameworks)

    # Domain overlap scored by Jaccard similarity.
    unit_domains = set(unit.domains)
    query_domain_set = set(query_domains)
    if unit_domains or query_domain_set:
        domain_score = len(unit_domains & query_domain_set) / len(unit_domains | query_domain_set)
    else:
        domain_score = 0.0

    # Language match: any overlap between query and unit languages.
    language_score = 0.0
    if query_languages and any(lang in unit.context.languages for lang in query_languages):
        language_score = 1.0

    # Framework match: any overlap between query and unit frameworks.
    framework_score = 0.0
    if query_frameworks and any(fw in unit.context.frameworks for fw in query_frameworks):
        framework_score = 1.0

    # Pattern match: exact case-insensitive equality between query and unit pattern.
    pattern_score = (
        1.0 if query_pattern and unit.context.pattern and query_pattern.lower() == unit.context.pattern.lower() else 0.0
    )

    score = (
        _DOMAIN_WEIGHT * domain_score
        + _LANGUAGE_WEIGHT * language_score
        + _FRAMEWORK_WEIGHT * framework_score
        + _PATTERN_WEIGHT * pattern_score
    )
    return min(max(score, _RELEVANCE_FLOOR), _RELEVANCE_CEILING)

File: sdk/go/scoring.go
package cq

import (
	"slices"
	"strings"
	"time"
)

// Scoring weights and confidence bounds used to rank knowledge units.
const (
	confidenceCeiling = 1.0
	confidenceFloor   = 0.0

	confirmationBoost = 0.1
	flagPenalty       = 0.15

	domainWeight    = 0.55
	frameworkWeight = 0.15
	languageWeight  = 0.15
	patternWeight   = 0.15
)

// relevance scores how relevant ku is to the given query parameters.
func (ku KnowledgeUnit) relevance(
	queryDomains []string,
	queryLanguages []string,
	queryFrameworks []string,
	queryPattern string,
) float64 {
	domainScore := jaccardSimilarity(ku.Domains, queryDomains)
	var languageScore, frameworkScore, patternScore float64
	if anyMatch(ku.Context.Languages, queryLanguages) {
		languageScore = 1.0
	}
	if anyMatch(ku.Context.Frameworks, queryFrameworks) {
		frameworkScore = 1.0
	}
	if queryPattern != "" && ku.Context.Pattern != "" && strings.EqualFold(queryPattern, ku.Context.Pattern) {
		patternScore = 1.0
	}
	score := domainWeight*domainScore + languageWeight*languageScore + frameworkWeight*frameworkScore + patternWeight*patternScore
	return min(max(score, confidenceFloor), confidenceCeiling)
}

// anyMatch reports whether any element in queries appears in items.
func anyMatch(items []string, queries []string) bool {
	if len(queries) == 0 {
		return false
	}
	set := make(map[string]struct{}, len(items))
	for _, item := range items {
		set[item] = struct{}{}
	}
	for _, q := range queries {
		if _, ok := set[q]; ok {
			return true
		}
	}
	return false
}

// applyConfirmation returns a copy of ku with boosted confidence and incremented confirmation count.
func applyConfirmation(ku KnowledgeUnit) KnowledgeUnit {
	out := ku
	out.Domains = slices.Clone(ku.Domains)
	out.Flags = slices.Clone(ku.Flags)
	out.Context.Languages = slices.Clone(ku.Context.Languages)
	out.Context.Frameworks = slices.Clone(ku.Context.Frameworks)
	out.Evidence.Confidence = min(out.Evidence.Confidence+confirmationBoost, confidenceCeiling)
	out.Evidence.Confirmations++
	now := time.Now()
	out.Evidence.LastConfirmed = &now
	return out
}

// applyFlag returns a copy of ku with reduced confidence and the given flag appended.
func applyFlag(ku KnowledgeUnit, reason FlagReason, cfg flagConfig) KnowledgeUnit {
	out := ku
	out.Domains = slices.Clone(ku.Domains)
	out.Context.Languages = slices.Clone(ku.Context.Languages)
	out.Context.Frameworks = slices.Clone(ku.Context.Frameworks)
	out.Evidence.Confidence = max(out.Evidence.Confidence-flagPenalty, confidenceFloor)
	now := time.Now()
	out.Flags = append(slices.Clone(ku.Flags), Flag{
		Reason:      reason,
		Timestamp:   &now,
		Detail:      cfg.detail,
		DuplicateOf: cfg.duplicateOf,
	})
	return out
}

// jaccardSimilarity computes the Jaccard index (intersection over union) of two string slices.
func jaccardSimilarity(a []string, b []string) float64 {
	if len(a) == 0 && len(b) == 0 {
		return 0.0
	}
	setA := make(map[string]struct{}, len(a))
	for _, v := range a {
		setA[v] = struct{}{}
	}
	setB := make(map[string]struct{}, len(b))
	for _, v := range b {
		setB[v] = struct{}{}
	}
	var intersection int
	for v := range setA {
		if _, ok := setB[v]; ok {
			intersection++
		}
	}
	union := len(setA) + len(setB) - intersection
	return float64(intersection) / float64(union)
}
